### DISCUSSION TITLE
Do you recommend any strategies for maps with multiple Z-levels?

### DISCUSSION THREAD

**Discussion by shanesveller** [StarArawn/bevy_ecs_tilemap#200] [Q&A]
I am working on a game that can be handwaved as a Dwarf Fortress clone and recently did a rewrite to introduce multiple Z-levels like DF has to represent depth above and below the surface of the larger world, permitting vertical building and navigation. Prior to this it would have played more similarly to Rimworld instead. I'm also currently evaluating [bevy_tileset](https://github.com/MrGVSV/bevy_tileset) and [bevy_tileset_map](https://github.com/MrGVSV/bevy_tileset_map).

My question here is intended to be limited specifically to integration with `bevy_ecs_tilemap` and not all of the many other knock-on changes that come with 2D->3D transition, since those are obviously my responsibility.

I currently am using a brute-force approach:
- Whenever I change my currently-visible-Z-level, stored in a Resource, I just clobber the entire `bevy_ecs_tilemap::Layer` and redraw it from scratch, for each of 3 layers (I currently separate terrain, mobiles, and indicators/overlays).
- If the Z-layer hasn't changed, I can do my usual incremental changes based on movement or events.

Do you or any of the other users of this library have any pro-tips to share? I'm somewhere around early-intermediate experience level with Rust, and my background is in non-systems languages like Elixir, so anything that seems obvious to a reader may not have occurred to me yet.

One nice-to-have thing I want to adapt from DF is that if you are looking at a Z-level that is sparsely populated, perhaps above the surface, level with the tip of a mountain/hill or tops of trees, you can see through unoccupied space to the layers below. I think I would need to introduce many more `Layers` and treat them more like a ringbuffer. I'm not sure if there's any efficiency to be gained through advice by someone knowledgeable of this crate's internals.

**Comment by IceSentry** [comment]
This is based on an old version of bevy and bevy_ecs_tilemap, but I did something similar in https://github.com/IceSentry/bevy_df it's been a year though, so I don't remember the details.

It's mostly contained here https://github.com/IceSentry/bevy_df/blob/master/src/map/renderer.rs

**Reply by IceSentry** [reply]
Here's an old gif of what it looked like

![ufRJXrtP9L](https://user-images.githubusercontent.com/8348954/180494235-b80384d3-b780-4dbb-913f-be1dae3562d2.gif)

**Reply by IceSentry** [reply]
I used isometric tiles, but the principles should be the same

**Reply by AskoldMakaruk** [reply]
Hey, that code depends on branch="improved-iso" which seems to be deleted. Can this be runned somehow without that?

**Reply by IceSentry** [reply]
I don't know, I haven't touch that in a while and I believe bevy_ecs_tilemap has had a pretty major refactor since then. The bevy version is probably also severely out of date.

**Comment by StarArawn** [comment]
I would highly suggest you ditch `bevy_ecs_tilemap` for 3d isometric. In order to render it efficiently you need to use a depth buffer but it presents a whole bunch of issues with bevy's default 2D rendering which has no depth buffer. I've been somewhat working on a game idea which involves a 3D isometric world. I wrote an iso renderer that uses a customer renderer for the map and sprites. Here is a picture:
![image](https://user-images.githubusercontent.com/6656977/180649220-faca2c6d-d248-4eb8-a83b-5df9fb746361.png)

I'm tempted to release a new crate called `bevy_iso` or something, but I need to finish up the rewrite here first.

**Reply by shanesveller** [reply]
I'm personally not doing isometric art style, just top-down tiles that abstractly convey a world with depth like DF does. I have no plans to change there.

**Reply by StarArawn** [reply]
Ideally you would use the re-write here, because it stores the data in a which provides a lot of flexibility. For a 2D DF like world I would spawn everything in and cull the layers which aren't visible. I would probably try to use some sort of chunking method as well as DF worlds tend to be quite massive.

**Reply by Affinator** [reply]
> I would highly suggest you ditch `bevy_ecs_tilemap` for 3d isometric. In order to render it efficiently you need to use a depth buffer but it presents a whole bunch of issues with bevy's default 2D rendering which has no depth buffer. I've been somewhat working on a game idea which involves a 3D isometric world. I wrote an iso renderer that uses a customer renderer for the map and sprites. Here is a picture: ![image](https://user-images.githubusercontent.com/6656977/180649220-faca2c6d-d248-4eb8-a83b-5df9fb746361.png)
> 
> I'm tempted to release a new crate called `bevy_iso` or something, but I need to finish up the rewrite here first.

Any news on this? After playing around with this crate I Unterstand the amount of work needed to get 2D with depth/height working...

**Reply by StarArawn** [reply]
> Any news on this? After playing around with this crate I Unterstand the amount of work needed to get 2D with depth/height working...

Sadly no, I just don't have the time to devote to this right now.

**Reply by Ygg01** [reply]
Are there any source on how to make this? If not a repo then a gist or something?

**Reply by StarArawn** [reply]
> Are there any source on how to make this? If not a repo then a gist or something?

Nope. I think a good starting point would be adding the option in bevy to include a depth buffer for bevy 2d. That might be something that's controversial though.
